Lib9c/Action/RuneEnhancement.cs

```diff
@@ -0,0 +1,155 @@
+            }
+            else
+            {
+                runeState = new RuneState(RuneId);
+            }
+
+            var costSheet = sheets.GetSheet<RuneCostSheet>();
+            if (!costSheet.TryGetValue(runeState.RuneId, out var costRow))
+            {
+                throw new RuneCostNotFoundException(
+                    $"[{nameof(RuneEnhancement)}] my avatar address : {AvatarAddress}");
+            }
+
+            var targetLevel = runeState.Level + 1;
+            if (!costRow.TryGetCost(targetLevel, out var cost))
+            {
+                throw new RuneCostDataNotFoundException(
+                    $"[{nameof(RuneEnhancement)}] my avatar address : {AvatarAddress}");
+            }
+
+            var runeSheet = sheets.GetSheet<RuneSheet>();
+            if (!runeSheet.TryGetValue(cost.RuneStoneId, out var runeRow))
+            {
+                throw new RuneNotFoundException(
+                    $"[{nameof(RuneEnhancement)}] my avatar address : {AvatarAddress}");
+            }
+
+            var ncgCurrency = states.GetGoldCurrency();
+            var crystalCurrency = CrystalCalculator.CRYSTAL;
+            var runeCurrency = Currency.Legacy(runeRow.Ticker, 0, minters: null);
+            var ncgBalance = states.GetBalance(context.Signer, ncgCurrency);
+            var crystalBalance = states.GetBalance(context.Signer, crystalCurrency);
+            var runeBalance = states.GetBalance(AvatarAddress, runeCurrency);
+            if (TryEnhancement(ncgBalance, crystalBalance, runeBalance,
+                    ncgCurrency, crystalCurrency, runeCurrency,
+                    cost, context.Random, Once, out var tryCount))
+            {
+                runeState.LevelUp();
+            }
+
+            var ncgCost = cost.NcgQuantity * tryCount * ncgCurrency;
+            var crystalCost = cost.CrystalQuantity * tryCount * crystalCurrency;
+            var runeCost = cost.RuneStoneQuantity * tryCount * runeCurrency;
+            var arenaSheet = sheets.GetSheet<ArenaSheet>();
+            var arenaData = arenaSheet.GetRoundByBlockIndex(context.BlockIndex);
+            var feeStoreAddress = Addresses.GetBlacksmithFeeAddress(arenaData.ChampionshipId, arenaData.Round);
+            return states.SetState(runeStateAddress, runeState.Serialize())
+                .TransferAsset(context.Signer, feeStoreAddress, ncgCost)
+                .TransferAsset(context.Signer, feeStoreAddress, crystalCost)
+                .TransferAsset(AvatarAddress, feeStoreAddress, runeCost);
+        }
+
+        private bool TryEnhancement(
+            FungibleAssetValue ncg,
+            FungibleAssetValue crystal,
+            FungibleAssetValue rune,
+            Currency ncgCurrency,
+            Currency crystalCurrency,
+            Currency runeCurrency,
+            RuneCostSheet.RuneCostData cost,
+            IRandom random,
+            bool once,
+            out int tryCount)
```

Comment:
it can be reusable in client side when action rendered. how about seperate external class like `RuneHelper`